server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/MockDiskUsagesIT.java

```diff
@@ -215,6 +221,87 @@ public void testAutomaticReleaseOfIndexBlock() throws Exception {
         assertSearchHits(client().prepareSearch("test").get(), "1", "3");
     }
 
+    public void testCancelExistingRecoveriesOnDiskPassingHighWatermark() throws Exception {
+
+        // start two data node
+        String node0 = internalCluster().startNode(Settings.builder()
+            .put(Environment.PATH_DATA_SETTING.getKey(), createTempDir())
+            .put("node.attr.tag", "node0").build());
+
+        String node1 = internalCluster().startNode(Settings.builder()
+            .put(Environment.PATH_DATA_SETTING.getKey(), createTempDir())
+            .put("node.attr.tag", "node1").build());
+
+        final MockInternalClusterInfoService clusterInfoService = getMockInternalClusterInfoService();
+        clusterInfoService.setUpdateFrequency(TimeValue.timeValueMillis(50));
+        clusterInfoService.onMaster();
+
+        ClusterState clusterState = client().admin().cluster().prepareState().get().getState();
+
+        // prevent any effects from in-flight recoveries, since we are only simulating a 100-byte disk
+        clusterInfoService.shardSizeFunction = shardRouting -> 0L;
+
+        // start with all nodes below the watermark
+        clusterInfoService.diskUsageFunction = (discoveryNode, fsInfoPath) -> setDiskUsage(fsInfoPath, 100, between(10, 100));
+
+        final boolean watermarkBytes = randomBoolean(); // we have to consistently use bytes or percentage for the disk watermark settings
+        assertAcked(client().admin().cluster().prepareUpdateSettings().setPersistentSettings(Settings.builder()
+            .put(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.getKey(), watermarkBytes ? "10b" : "90%")
+            .put(CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey(), watermarkBytes ? "10b" : "90%")
+            .put(CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey(), watermarkBytes ? "0b" : "100%")
+            .put(CLUSTER_ROUTING_ALLOCATION_REROUTE_INTERVAL_SETTING.getKey(), "1s")
+            .put(INDICES_RECOVERY_INTERNAL_ACTION_TIMEOUT_SETTING.getKey(), "10s")
+            .put(INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING.getKey(), "1b")));
+        // Create an index with 2 shards with 1 replica, but only allow allocation on node0, so all replicas are left unassigned
+        final String index = "test";
+        assertAcked(prepareCreate(index).setSettings(Settings.builder()
+            .put("number_of_shards", 2)
```

Comment:
Should work with any number of shards:

```suggestion
            .put("number_of_shards", between(1, 5))
```